app/models/concerns/due_date_permissions.rb

```diff
@@ -0,0 +1,199 @@
+
+  def metareview_permissible?
+    activity_permissible?(:metareview) if respond_to?(:metareview_allowed_id)
+  end
+
+  # Check if deadline allows late submissions
+  def allows_late_submission?
+    return false unless submission_allowed_id
+
+    deadline_right = DeadlineRight.find_by(id: submission_allowed_id)
+    deadline_right&.name == 'Late'
+  end
+
+  def allows_late_review?
+    return false unless review_allowed_id
+
+    deadline_right = DeadlineRight.find_by(id: review_allowed_id)
+    deadline_right&.name == 'Late'
+  end
+
+  def allows_late_quiz?
+    return false unless quiz_allowed_id
+
+    deadline_right = DeadlineRight.find_by(id: quiz_allowed_id)
+    deadline_right&.name == 'Late'
+  end
+
+  # Check if any activity is currently allowed
+  def has_any_permission?
+    can_submit? || can_review? || can_take_quiz? || can_teammate_review? ||
+      (respond_to?(:can_metareview?) && can_metareview?)
+  end
+
+  # Get list of currently allowed activities
+  def allowed_activities
+    activities = []
+    activities << 'submission' if can_submit?
+    activities << 'review' if can_review?
+    activities << 'quiz' if can_take_quiz?
+    activities << 'teammate_review' if can_teammate_review?
+    activities << 'metareview' if respond_to?(:can_metareview?) && can_metareview?
+    activities
+  end
+
+  # Check if this deadline is currently active (allows some action)
+  def active?
+    has_any_permission?
+  end
+
+  # Check if this deadline is completely closed (no actions allowed)
+  def closed?
+    !active?
+  end
+
+  # Check permissions for deadline type compatibility
+  def deadline_type_permits_action?(action)
+    return false unless deadline_type
+
+    case action.to_s.downcase
+    when 'submit', 'submission'
+      deadline_type.allows_submission?
+    when 'review'
+      deadline_type.allows_review?
+    when 'quiz'
+      deadline_type.allows_quiz?
+    when 'teammate_review'
+      deadline_type.allows_review?
+    when 'metareview'
+      deadline_type.allows_review?
+    else
+      false
+    end
+  end
+
+  # Comprehensive permission check combining deadline type and deadline rights
+  def permits_action?(action)
+    deadline_type_permits_action?(action) && activity_permissible?(action)
+  end
+
+  # Get permission status for an action (OK, Late, No)
+  def permission_status_for(action)
+    permission_field = "#{action}_allowed_id"
+    return 'No' unless respond_to?(permission_field)
+
+    allowed_id = public_send(permission_field)
+    return 'No' unless allowed_id
+
+    deadline_right = DeadlineRight.find_by(id: allowed_id)
+    deadline_right&.name || 'No'
+  end
+
+  # Check if deadline is in grace period (allows late submissions)
+  def in_grace_period_for?(action)
+    permission_status_for(action) == 'Late'
+  end
+
+  # Check if deadline is fully open for action
+  def fully_open_for?(action)
+    permission_status_for(action) == 'OK'
+  end
+
+  # Get human-readable permission description
```

Comment:
I'm open to having this method, but we should think carefully of where it might be called from, and how the message should be worded.